caffe2/core/typeid.h

```diff
@@ -352,25 +379,39 @@ class TypeMeta {
 // and as a result, we define these two macros slightly differently.
 
 #ifdef _MSC_VER
-#define CAFFE_KNOWN_TYPE(T)                              \
-  template <>                                            \
-  CAFFE2_EXPORT CaffeTypeId TypeMeta::Id<T>() {          \
-    static bool type_id_bit[1];                          \
-    static TypeNameRegisterer<T> registerer(             \
-        reinterpret_cast<CaffeTypeId>(type_id_bit), #T); \
-    return reinterpret_cast<CaffeTypeId>(type_id_bit);   \
+#define C10_KNOWN_TYPE(T)                                                     \
+  template <>                                                                 \
+  CAFFE2_EXPORT TypeId TypeMeta::Id<T>() {                                    \
+    static bool type_id_bit[1];                                               \
+    static const TypeId type_id(reinterpret_cast<intptr_t>(type_id_bit));     \
+    static TypeNameRegisterer<T> registerer(                                  \
+        type_id, #T);                                                         \
+    return type_id;                                                           \
   }
 #else // _MSC_VER
-#define CAFFE_KNOWN_TYPE(T)                              \
-  template <>                                            \
-  CaffeTypeId TypeMeta::Id<T>() {                        \
-    static bool type_id_bit[1];                          \
-    static TypeNameRegisterer<T> registerer(             \
-        reinterpret_cast<CaffeTypeId>(type_id_bit), #T); \
-    return reinterpret_cast<CaffeTypeId>(type_id_bit);   \
+#define C10_KNOWN_TYPE(T)                                                     \
+  template <>                                                                 \
+  TypeId TypeMeta::Id<T>() {                                                  \
+      static bool type_id_bit[1];                                             \
+    static const TypeId type_id(reinterpret_cast<intptr_t>(type_id_bit));     \
+    static TypeNameRegisterer<T> registerer(                                  \
+        type_id, #T);                                                         \
+    return type_id;                                                           \
   }
 #endif
 
-} // namespace caffe2
+}
+
+// Define adapters for legacy code
+namespace caffe2 {
+using CaffeTypeId = c10::TypeId;
+using TypeMeta = c10::TypeMeta;
+
+// Needs to be called from ::caffe2 namespace
+#define CAFFE_KNOWN_TYPE(T)                              \
+  } namespace c10 {                                      \
```

Comment:
This looks... a little dangerous since it is macros manipulating namespace :)

For the sake of simplicity, can we choose CAFFE_KNOWN_TYPE or C10_KNOWN_TYPE and unify all call sites, since we are in the process of migration and there probably wouldn't be a better time?

Reply:
The `CAFFE_KNOWN_TYPE` already has the hard requirement that it's called inside the `::caffe2` namespace because c++ compilers only allow specializing when you're in the same namespace as the original definition. This doesn't change this requirement.

In the end, the goal would be to have `C10_KNOWN_TYPE` everywhere, but it's not my highest priority currently.

Reply:
There's a difference: in the old implementation, the specialization would just fail if you put it in the wrong namespace. In this version, specialization will still succeed, and you'll just have everything after `CAFFE_KNOWN_TYPE` put in the wrong namespace. This will be **very confusing** for the first person to trip up on this. If you can somehow assert that you were previously in the caffe2 namespace I will feel much better about this.

Reply:
That's why I'd say to put it on the top namespace or just not change it all.

Reply:
OK, after reading the codebase I do have a strong preference to keep the CAFFE2_KNOWN_TYPE, for the technical reason that - like @ezyang said - it is **very confusing** to users if things go wrong and it is very error prone. Let's make sure that we are not manipulating namespaces.